components/tooltip/tooltip.js

```diff
@@ -0,0 +1,238 @@
+import { clearDismissible, setDismissible } from '../../helpers/dismissible.js';
+import { css, html, LitElement } from 'lit-element/lit-element.js';
+import { bodyCompactStyles } from '../typography/styles.js';
+import { getUniqueId } from '../../helpers/uniqueId.js';
+import { RtlMixin } from '../../mixins/rtl-mixin.js';
+
+class Tooltip extends RtlMixin(LitElement) {
+
+	static get properties() {
+		return {
+			delay: { type: Number },
+			disableFocusLock: { type: Boolean, attribute: 'disable-focus-lock' },
+			for: { type: String },
+			forceShow: { type: Boolean, attribute: 'force-show' },
+			showing: { type: Boolean, reflect: true, attribute: 'showing' },
+			state: { type: String, reflect: true }, /* Valid values are: 'info' and 'error' */
+		};
+	}
+
+	static get styles() {
+		return [bodyCompactStyles, css`
+			:host {
+				--d2l-tooltip-background-color: var(--d2l-color-ferrite); /* Deprecated, use state attribute instead */
+				background-color: var(--d2l-tooltip-background-color);
+				box-sizing: border-box;
+				color: white;
+				display: none;
+				position: absolute;
+				text-align: left;
+				white-space: normal;
+				z-index: 1000; /* position on top of floating buttons */
+			}
+
+			:host([state="error"]) {
+				--d2l-tooltip-background-color: var(--d2l-color-cinnabar);
+			}
+
+			:host([dir="rtl"]) {
+				text-align: right;
+			}
+
+			:host([showing]) {
+				display: inline-block;
+			}
+		`];
+	}
+
+	constructor() {
+		super();
+
+		this._onMouseEnter = this._onMouseEnter.bind(this);
+		this._onMouseLeave = this._onMouseLeave.bind(this);
+		this._onFocus = this._onFocus.bind(this);
+		this._onBlur = this._onBlur.bind(this);
+
+		this.delay = 0;
+		this.disableFocusLock = false;
+		this.forceShow = false;
+		this.showing = false;
+		this.state = 'info';
+
+		this._isFocusing = false;
+		this._isHovering = false;
+		this._dismissibleId = null;
+	}
+
+	get for() {
+		return this._for;
+	}
+	set for(val) {
+		const oldVal = this._for;
+		if (oldVal !== val) {
+			this._for = val;
+			this.requestUpdate('for', oldVal);
+			this._updateTarget();
+		}
```

Comment:
I wonder if we need the accessor or if just using the `attributeChangedCallback` and checking for it in there to trigger `_updateTarget` would work? Same for `forceShow` and `showing`.

Reply:
That'll definitely work. For my own understanding, is there a reason to prefer one versus the other?

I was leaning towards using `setters` / `getters` since they'll only run when that attribute changes versus `attributeChangedCallback` which runs whenever any attribute changes and needs to explicitly check each time to see what the name of the changed attribute was to determine which code block to run.

Reply:
I don't have enough experience with them to have a strong opinion. I think setters/getters are obviously great if you need to validate/normalize the value before actually letting it get through. But yeah, don't feel strongly -- good with the getters/setters here.

Reply:
I've also been contemplating this a bit.  I've also found `update(changedProperties)` to be worthwhile - it gets called right after `attributeChangedCallback`, before `render`.